redis/src/cluster_async/mod.rs

```diff
@@ -195,18 +304,116 @@ type ConnectionMap<C> = connections_container::ConnectionsMap<ConnectionFuture<C
 type ConnectionsContainer<C> =
     self::connections_container::ConnectionsContainer<ConnectionFuture<C>>;
 
-struct InnerCore<C> {
-    conn_lock: RwLock<ConnectionsContainer<C>>,
+pub(crate) struct InnerCore<C> {
+    pub(crate) conn_lock: RwLock<ConnectionsContainer<C>>,
     cluster_params: ClusterParams,
     pending_requests: Mutex<Vec<PendingRequest<C>>>,
-    slot_refresh_in_progress: AtomicBool,
+    pub(crate) slot_refresh_in_progress: AtomicBool,
     initial_nodes: Vec<ConnectionInfo>,
 }
 
-type Core<C> = Arc<InnerCore<C>>;
+pub(crate) type Core<C> = Arc<InnerCore<C>>;
+
+impl<C> InnerCore<C>
+where
+    C: ConnectionLike + Connect + Clone + Send + Sync + 'static,
+{
+    // Check if all slots are covered by the cluster
+    pub(crate) async fn all_slots_covered(&self) -> bool {
+        let conn = self.conn_lock.read().await;
+        conn.slot_map.is_all_slots_covered()
+    }
 
-struct ClusterConnInner<C> {
-    inner: Core<C>,
+    // return address of node for slot
+    pub(crate) async fn get_address_from_slot(
+        &self,
+        slot: u16,
+        slot_addr: SlotAddr,
+    ) -> Option<String> {
+        self.conn_lock
+            .read()
+            .await
+            .slot_map
+            .get_node_address_for_slot(slot, slot_addr)
+    }
+
+    // return epoch of node
+    pub(crate) async fn get_address_epoch(&self, node_address: &str) -> Result<u64, RedisError> {
+        let command = cmd("CLUSTER").arg("INFO").to_owned();
+        let node_conn = self
+            .conn_lock
+            .read()
+            .await
+            .connection_for_address(node_address);
```

Comment:
Nit: you could simplify this by using `ok_or` to convert the None into an Err and then use the `?` operator


```rust
et node_conn = self
            .conn_lock
            .read()
            .await
            .connection_for_address(node_address).ok_or(RedisError::from((
                                ErrorKind::ResponseError,
                                "Failed to parse cluster info",
                            ))?;
```